### ISSUE TITLE
Voicemail to Email

### ISSUE THREAD

**Issue by udy786** [fusionpbx/fusionpbx#2567]
Hi,

I am using FusionPBX 4.2.0 on debian. Switch 1.6.15 (64bit). When getting voicemail then extension users not getting voicemail on email. In Status -> Emails, not getting any failed logs. Configured SMTP in Advance -> Default Settings -> Email. With same SMPT details email working on other server.  

Extension setting:-
Voicemail Enabled: True
Voicemail Mail To:- UserMailID
Voicemail Keep Local:-Ture

Voicemail showing inside App -> Voicemail of each user. From where able to play, Download and Delete voicemail. 

Checked logs in /tmp but not getting any log file or error logs. 

On CLI its showing like 

[DEBUG] switch_utils.c:1180 Emailed file [/tmp/mail.1494243604cfe8] to [user@email.com]

Did Google and Followed all possible documents and URL but but same result that not working. 

Please help.

**Comment by tonyhain** [comment]
I saw the same behavior on 4.2.2 on FreeBSD, but hadn't worried about it yet due to higher priorities. The mta's in the MX records never show any connection attempts, so I am at a loss as to where it thinks it delivered the mail. In any case, this would appear to be a FreeSwitch issue not specifically related to FusionPBX. When I get time to worry about that issue, i will run tcpdump when the voicemail is recorded to see where exactly the FS mail service is trying to send it.

**Comment by powerpbx** [comment]
If you don't have fusionpbx installed in `/var/www/fusionpbx` then you need to change that mailer-app value in `/etc/freeswitch/autoload_configs/switch.conf.xml`

**Comment by tonyhain** [comment]
Thanks for the pointer, because it ended up in /usr/local/www here. Too many hidden little assumptions about directory structures. The only thing consistent about all the flavors of *nix over the last 40 years is that there is zero consistency. 
Assuming fusionpbx knows where it is installed, it should check that line the first time someone sets an email address in the box. In any case, freeswitch should not claim in the log that email was sent when there was no way it could have found the mailer.

**Comment by markjcrane** [comment]
There is many ways to do this command will help you find the file.
find / -name switch.conf.xml
You can modify the file to use freeswitch default if you would like and setup sendmail, postfix or ssmtp.

**Comment by udy786** [comment]
Hello powerpbx, Fusionpbx installed in /var/www/fusionpbx. I am able to access user interface from browser. 
Here is value of /etc/freeswitch/autoload_configs/switch.conf.xml
 <param name="mailer-app" value="/usr/bin/php /var/www/fusionpbx/secure/v_mailto.php"/>

Also I checked for /var/www/fusionpbx/secure/v_mailto.php and file exist on this location.

**Comment by tonyhain** [comment]
My issue was not finding the file, it was knowing that there was a pointer in it that didn't know where fusionpbx was installed. The UI options in fusionpbx for email/smtp & default settings don't mention that FS was told to send email to a non-existent path. If that was set by fusionpbx during install, the script should have recognized where fusionpbx was actually installed. Why would an end user know something needs to be fixed when the UI offers email settings, but not that one?

**Comment by udy786** [comment]
Hello tonyhain, 

I would like to request you, if you problem same as I posted or related to voicemail not going to email then only please include here else please create new issue. We are creating confusion for everyone to reply.

**Comment by tonyhain** [comment]
udy786,  This is about voicemail to email failing in the same way you described. 
I edited the switch.conf.xml to point to /usr/local/www/fusionpbx, restarted FS, and like yours the FS log shows:
2017-05-09 21:37:36.079930 [DEBUG] switch_utils.c:1180 Emailed file [/tmp/mail.14943910568154] to ...
Default Settings - Email  : method smtp  : smtp host  smtp.tndh.net
ping smtp.tndh.net 
PING express.tndh.net (172.20.144.31): 56 data bytes
64 bytes from 172.20.144.31: icmp_seq=0 ttl=63 time=0.830 ms
(so the name resolves and is reachable from the FS/fusionpbx system)

# tcpdump port 25 or port 587   on FS   ::: shows absolutely nothing happens

time was aligned so logs match:::   # ntpdate smtp.tndh.net   (cname is ntp2.tndh.net)
9 May 21:49:45 ntpdate[22465]: adjust time server 172.20.144.31 offset 0.009474 sec

# grep "2017-05-09 21:3" /var/log/exim/mainlog
2017-05-09 21:33:47 H=vds2026.sivit.org [195.5.208.252] ...: SPAMHAUS match ...
2017-05-09 21:37:37 1d8JMw-0001Kz-G2 ... H=mail.bcofsa.com.ar [200.45.255.76] ... rejected message: score=5.7 ...
2017-05-09 21:38:11 1d8JNz-0001Ly-U7 H=(host.textilesanahuac.info) [67.227.245.7] ... rejected after DATA: This message scored 11.8 spam points.
2017-05-09 21:38:39 H=(thirdlane.local) [209.151.146.9] ...: SPAMHAUS match ...

Basically, someone is lying. FS claims the email was sent, but tcpdump and the exim log say nothing was even attempted. It is still not clear to me if this is a FS problem or fusionpbx, but given the need to edit the xml file, it is looking like FS may not be getting the proper information from fusionpbx. Either way the log should flag an error rather than claim email was sent when the xml file was pointing at a non-existent path.

**Comment by udy786** [comment]
Today I changed mailer-app to sendmail and now I am getting voicemail on email. Email getting all html code in body. Anyone please help to find template file location? I tried to modify /etc/freeswitch/voicemail.tpl no effect. Looks like Fusionpbx using other template file from other location. 

@tonyhain you should try to set mailer-app to sendmail. 

When I am using mailer-app to /usr/bin/php /var/www/fusionpbx/secure/v_mailto.php then only its not working. I tried to run v_mailto.php manually and getting error 

PHP Fatal error:  Allowed memory size of 262144 bytes exhausted (tried to allocate 49152 bytes) in /var/www/fusionpbx/secure/v_mailto.php on line 151

@tonyhain Can you please try to run v_mailto.php manually by running below command. 

php /var/www/fusionpbx/secure/v_mailto.php

**Comment by udy786** [comment]
Finally my problem solved. I modified php.ini on server to increase memory limit. This file was located on two different location. 

/etc/php5/fpm/php.ini
/etc/php5/cli/php.ini

**Comment by tonyhain** [comment]
One of the first things my build scripts do is disable sendmail and replace it with exim. In this case though I hadn't run the full build since I am still trying to figure out the script for fusionpbx/FS, so I changed to sendmail in the fusionpbx UI with the same result. Note: the switch.conf.xml file still shows the mailer app as v_mailto.php

v_mailto.php doesn't do anything ... before or after changing smtp to sendmail
# ls -l /usr/local/www/fusionpbx/secure/v_mailto.php
-rwxr-xr-x  1 root  www  12845 Apr 26 19:38 /usr/local/www/fusionpbx/secure/v_mailto.php
# php /usr/local/www/fusionpbx/secure/v_mailto.php
^C
# 
--- in another ssh session
# tcpdump port 25 or port 587
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on hn0, link-type EN10MB (Ethernet), capture size 262144 bytes
^C
0 packets captured

**Comment by udy786** [comment]
You change sendmail  from switch.conf.xml and also when you run below command it should give you some output not blank. 
php /usr/local/www/fusionpbx/secure/v_mailto.php

It mean when Freeswitch calling mailer-app then not getting any reply or response. You will get any request on port and log in tcpdump till the time code will not run so first try run file v_mailto.php and get output or change sendmail  from switch.conf.xml. Don't forget to restart Freeswitch after doing any changes in  switch.conf.xml.

**Comment by tonyhain** [comment]
So getting it to work by bypassing the fusionpbx mailer pretty much isolates it to a fusionpbx problem. That would be consistent with FS believing it had handed off the item to the mailer-app so it is valid to mark it as sent, and any problem is an issue for the mailer-app. 

I don't see a way to cause v_mailto.php to log anything to show its progress. (no logging for any fusionpbx actions as far as that goes) There is a line in the file about //save output to
	$fp = fopen(sys_get_temp_dir()."/mailer-app.log", "w");
but that doesn't appear to be a real log file so much as a temp buffer with an unfortunate name.
The segment above that explains the hang though:
//listen for standard input
        if ($msg == '') {
                $fd = fopen("php://stdin", "r");
....
since I didn't give it a file, it is waiting for me to type the header. Actually, putting garbage in stdin does cause the above "log file" to leave a trace of the errors, so it is not a transaction log, just an error report. In any case, the log file didn't exist before the junk input, so it was either never created through the FS process, or fusionpbx thought the process completed successfully and deleted it. 

Well thinking about that statement for a second caused me to to revisit the point from powerpbx earlier. I had changed the prefix from /var/www to /usr/local/www, but didn't change the one for php itself.      which php
/usr/local/bin/php
Once the entry in switch.conf.xml included both the php executable and the path to fusionpbx mailer, it started sending mail.

**Comment by powerpbx** [comment]
Log file is located at /tmp/mailer-app.log

**Comment by tonyhain** [comment]
I should have thought of it earlier. Looking to v_mailto.php I realized that if it had ever been run there would be a log file, but there wasn't until I ran php by hand. That meant php was not on the specified path, which proved to be the case. 
Since email was just a symptom of the overall issue here, I opened a new issue about the core point of path mismatch. There were others I found in a quick search, but there might be some I didn't know enough to know where to look.

FWIW: lately when I write cross-platform scripts I load a variable for executables using 'which' as a first try, but even that has failed at times. When all else fails:
find -f / | grep bin\/php$
;)
